app/lib/csv_importer.rb

```diff
@@ -10,29 +10,49 @@ class CsvImporter
 
   class InvalidCategoryError < StandardError; end;
 
-  def self.import(filename, category_name, processed_file_id)
-    importer = new(filename, category_name, processed_file_id)
-    importer.process
-    importer.stats
-  end
-
   def initialize(filename, category_name, processed_file_id)
     @filename = filename
     @processed_file_id = processed_file_id
     @model = model_from_category(category_name)
-
     @stats = Hash.new(0)
+    @error_details = {}
     @stats[:shl_case_numbers] = Hash.new(0)
   end
 
+  def call
+    process
+  end
+
+  def errored?
+    !error_details.empty?
+  end
+
+  private
+
   def process
-    IOStreams.each_record(filename) do |csv_row|
-      csv_row[:processed_file_id] = processed_file_id
-      csv_row[:raw] = false
-      record = model.create_from_csv_data(csv_row)
-      record.cleanse_data! if record.respond_to?(:cleanse_data!)
-      record.save
-      increment_stats(record)
+    row_number = 2 # assuming 1 is headers
+
+    model.transaction do
+      IOStreams.each_record(filename) do |csv_row|
+        csv_row[:processed_file_id] = processed_file_id
+        csv_row[:raw] = false
+        record = model.create_from_csv_data(csv_row)
+        record.cleanse_data! if record.respond_to?(:cleanse_data!)
+
+        unless record.valid?
+          error_details["row_number_#{row_number}"] = record.errors.details
+        else
+          record.save
+          increment_stats(record)
+        end
```

Comment:
https://github.com/rubocop-hq/ruby-style-guide#no-else-with-unless

Reply:
Nice catch @rafaltrojanowski!

If you're so inclined, it'd be super helpful if you could help get style catches automated:
https://github.com/rubyforgood/abalone/issues/81 (previous WIP on this here: https://github.com/rubyforgood/abalone/pull/90)

Reply:
used `if...else` instead. Still trying to get used to RoR best practices 😄  Thanks for letting me know.